File: server/lib/channels/m3u.ts
import { buildChannelIdentity } from './identity'
import type { Channel, ParseM3UOptions } from './types'

type PendingChannel = {
  tvgId?: string
  number?: string
  name: string
  logoUrl?: string
  groupTitle?: string
  lineNumber: number
}

type ParsedExtInf = Omit<PendingChannel, 'lineNumber'>
type ParsedChannel = Channel & {
  lineNumber: number
}

const ALLOWED_SCHEMES = new Set(['http:', 'https:'])

function hasNumericGuideNumber(value: string | undefined): value is string {
  return value !== undefined && /^\d+$/.test(value)
}

function compareCaseInsensitive(left: string, right: string): number {
  const leftFolded = left.toLowerCase()
  const rightFolded = right.toLowerCase()

  if (leftFolded < rightFolded) {
    return -1
  }

  if (leftFolded > rightFolded) {
    return 1
  }

  return 0
}

function compareParsedChannels(left: ParsedChannel, right: ParsedChannel): number {
  const leftHasNumber = hasNumericGuideNumber(left.number)
  const rightHasNumber = hasNumericGuideNumber(right.number)

  if (leftHasNumber && rightHasNumber) {
    const numberDiff = Number(left.number) - Number(right.number)
    if (numberDiff !== 0) {
      return numberDiff
    }

    return compareCaseInsensitive(left.identity.key, right.identity.key)
  }

  if (leftHasNumber !== rightHasNumber) {
    return leftHasNumber ? -1 : 1
  }

  const nameDiff = compareCaseInsensitive(left.name, right.name)
  if (nameDiff !== 0) {
    return nameDiff
  }

  return compareCaseInsensitive(left.identity.key, right.identity.key)
}

function parseAttributes(value: string): Record<string, string> {
  const attributes: Record<string, string> = {}
  const pattern = /([A-Za-z0-9_-]+)="([^"]*)"/g
  let match: RegExpExecArray | null

  while ((match = pattern.exec(value)) !== null) {
    attributes[match[1].toLowerCase()] = match[2]
  }

  return attributes
}

function findExtInfSeparator(line: string): number {
  let inQuotes = false
  for (let index = '#EXTINF:'.length; index < line.length; index += 1) {
    const character = line[index]
    if (character === '"') {
      inQuotes = !inQuotes
      continue
    }

    if (character === ',' && !inQuotes) {
      return index
    }
  }

  return -1
}

function parseExtInf(line: string): ParsedExtInf | null {
  if (!line.startsWith('#EXTINF:')) {
    return null
  }

  const separatorIndex = findExtInfSeparator(line)
  if (separatorIndex < 0) {
    return null
  }

  const header = line.slice(0, separatorIndex)
  const name = line.slice(separatorIndex + 1).trim()
  if (!name) {
    return null
  }

  const attributes = parseAttributes(header)
  return {
    tvgId: attributes['tvg-id'],
    number: attributes['tvg-chno'],
    name,
    logoUrl: attributes['tvg-logo'],
    groupTitle: attributes['group-title']
  }
}

function isAllowedStreamUrl(value: string): boolean {
  try {
    const url = new URL(value)
    return ALLOWED_SCHEMES.has(url.protocol)
  } catch {
    return false
  }
}

function buildChannel(pending: PendingChannel, streamUrl: string): Channel {
  const identity = buildChannelIdentity({
    tvgId: pending.tvgId,
    number: pending.number,
    name: pending.name,
    streamUrl
  })

  return {
    identity,
    tvgId: pending.tvgId,
    number: pending.number,
    name: pending.name,
    logoUrl: pending.logoUrl,
    groupTitle: pending.groupTitle,
    streamUrl
  }
}

export function parseM3U(playlist: string, options: ParseM3UOptions): Channel[] {
  const channels: ParsedChannel[] = []
  let pending: PendingChannel | null = null

  const finalizePending = (streamUrl: string): void => {
    if (!pending) {
      return
    }

    const channel = buildChannel(pending, streamUrl)
    channels.push({
      ...channel,
      lineNumber: pending.lineNumber
    })

    pending = null
  }

  const dropPending = (reason: string): void => {
    if (!pending) {
      return
    }

    options.logger.warn(`dropped invalid channel at line ${pending.lineNumber}: ${reason}`)
    pending = null
  }

  const lines = playlist.split(/\r?\n/)
  for (let index = 0; index < lines.length; index += 1) {
    const lineNumber = index + 1
    const rawLine = lines[index]
    const line = rawLine.trim()

    if (!line || line === '#EXTM3U') {
      continue
    }

    const extInf = parseExtInf(line)
    if (extInf) {
      dropPending('missing stream url')
      pending = { ...extInf, lineNumber }
      continue
    }

    if (line.startsWith('#EXTINF:')) {
      options.logger.warn(`dropped invalid channel at line ${lineNumber}: malformed metadata`)
      continue
    }

    if (line.startsWith('#')) {
      continue
    }

    if (!isAllowedStreamUrl(rawLine)) {
      if (pending) {
        dropPending('unsupported or invalid stream url')
      } else {
        options.logger.warn(`dropped invalid channel at line ${lineNumber}: unexpected stream url`)
      }
      continue
    }

    if (!pending) {
      options.logger.warn(`dropped invalid channel at line ${lineNumber}: stream url without metadata`)
      continue
    }

    finalizePending(rawLine)
  }

  dropPending('missing stream url')

  channels.sort(compareParsedChannels)

  const orderedChannels: Channel[] = []
  const seen = new Set<string>()
  for (const channel of channels) {
    if (seen.has(channel.identity.key)) {
      options.logger.warn(`dropped duplicate channel at line ${channel.lineNumber}: ${channel.identity.key}`)
      continue
    }

    seen.add(channel.identity.key)
    orderedChannels.push({
      identity: channel.identity,
      tvgId: channel.tvgId,
      number: channel.number,
      name: channel.name,
      logoUrl: channel.logoUrl,
      groupTitle: channel.groupTitle,
      streamUrl: channel.streamUrl
    })
  }

  return orderedChannels
}

File: internal/channels/m3u.go
package channels

import (
	"fmt"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

type WarnLogger interface {
	Warn(message string, context any)
}

var (
	attributePattern   = regexp.MustCompile(`([A-Za-z0-9_-]+)="([^"]*)"`)
	allowedSchemes     = map[string]struct{}{"http": {}, "https": {}, "rtsp": {}, "rtsps": {}}
	numericGuideNumber = regexp.MustCompile(`^\d+$`)
)

type pendingChannel struct {
	TVGID      string
	Number     string
	Name       string
	LogoURL    string
	GroupTitle string
	LineNumber int
}

func ParseM3U(playlist string, logger WarnLogger) []Channel {
	lines := strings.Split(playlist, "\n")
	channels := make([]Channel, 0)
	var pending *pendingChannel
	sourceIndex := 0

	dropPending := func(reason string) {
		if pending == nil {
			return
		}
		logger.Warn(fmt.Sprintf("dropped invalid channel at line %d: %s", pending.LineNumber, reason), nil)
		pending = nil
	}

	finalizePending := func(streamURL string) {
		if pending == nil {
			return
		}
		identity := BuildChannelIdentity(pending.TVGID, pending.Number, pending.Name, streamURL)
		channels = append(channels, Channel{
			ID:          BuildChannelID(identity.Key),
			SourceIndex: sourceIndex,
			Identity:    identity,
			TVGID:       pending.TVGID,
			Number:      pending.Number,
			Name:        pending.Name,
			LogoURL:     pending.LogoURL,
			GroupTitle:  pending.GroupTitle,
			StreamURL:   streamURL,
		})
		sourceIndex++
		pending = nil
	}

	for index, rawLine := range lines {
		lineNumber := index + 1
		line := strings.TrimSpace(rawLine)
		if line == "" || line == "#EXTM3U" {
			continue
		}

		if extInf, ok := parseExtInf(line); ok {
			dropPending("missing stream url")
			pending = &pendingChannel{
				TVGID:      extInf.TVGID,
				Number:     extInf.Number,
				Name:       extInf.Name,
				LogoURL:    extInf.LogoURL,
				GroupTitle: extInf.GroupTitle,
				LineNumber: lineNumber,
			}
			continue
		}

		if strings.HasPrefix(line, "#EXTINF:") {
			logger.Warn(fmt.Sprintf("dropped invalid channel at line %d: malformed metadata", lineNumber), nil)
			continue
		}

		if strings.HasPrefix(line, "#") {
			continue
		}

		if !isAllowedStreamURL(rawLine) {
			if pending != nil {
				dropPending("unsupported or invalid stream url")
			} else {
				logger.Warn(fmt.Sprintf("dropped invalid channel at line %d: unexpected stream url", lineNumber), nil)
			}
			continue
		}

		if pending == nil {
			logger.Warn(fmt.Sprintf("dropped invalid channel at line %d: stream url without metadata", lineNumber), nil)
			continue
		}

		finalizePending(rawLine)
	}

	dropPending("missing stream url")

	sort.Slice(channels, func(i, j int) bool {
		return compareChannels(channels[i], channels[j]) < 0
	})

	ordered := make([]Channel, 0, len(channels))
	seen := map[string]struct{}{}
	for _, channel := range channels {
		if _, exists := seen[channel.Identity.Key]; exists {
			logger.Warn(fmt.Sprintf("dropped duplicate channel: %s", channel.Identity.Key), nil)
			continue
		}
		seen[channel.Identity.Key] = struct{}{}
		ordered = append(ordered, channel)
	}

	return ordered
}

type parsedExtInf struct {
	TVGID      string
	Number     string
	Name       string
	LogoURL    string
	GroupTitle string
}

func parseExtInf(line string) (parsedExtInf, bool) {
	if !strings.HasPrefix(line, "#EXTINF:") {
		return parsedExtInf{}, false
	}

	separatorIndex := findExtInfSeparator(line)
	if separatorIndex < 0 {
		return parsedExtInf{}, false
	}

	header := line[:separatorIndex]
	name := strings.TrimSpace(line[separatorIndex+1:])
	if name == "" {
		return parsedExtInf{}, false
	}

	attributes := map[string]string{}
	for _, match := range attributePattern.FindAllStringSubmatch(header, -1) {
		attributes[strings.ToLower(match[1])] = match[2]
	}

	return parsedExtInf{
		TVGID:      attributes["tvg-id"],
		Number:     attributes["tvg-chno"],
		Name:       name,
		LogoURL:    attributes["tvg-logo"],
		GroupTitle: attributes["group-title"],
	}, true
}

func findExtInfSeparator(line string) int {
	inQuotes := false
	for index := len("#EXTINF:"); index < len(line); index++ {
		switch line[index] {
		case '"':
			inQuotes = !inQuotes
		case ',':
			if !inQuotes {
				return index
			}
		}
	}
	return -1
}

func isAllowedStreamURL(value string) bool {
	parsed, err := url.Parse(value)
	if err != nil || parsed.Scheme == "" {
		return false
	}
	_, ok := allowedSchemes[parsed.Scheme]
	return ok
}

func compareChannels(left, right Channel) int {
	leftHasNumber := numericGuideNumber.MatchString(left.Number)
	rightHasNumber := numericGuideNumber.MatchString(right.Number)

	if leftHasNumber && rightHasNumber {
		leftNumber, _ := strconv.Atoi(left.Number)
		rightNumber, _ := strconv.Atoi(right.Number)
		if leftNumber != rightNumber {
			if leftNumber < rightNumber {
				return -1
			}
			return 1
		}
	}

	if leftHasNumber != rightHasNumber {
		if leftHasNumber {
			return -1
		}
		return 1
	}

	leftName := strings.ToLower(left.Name)
	rightName := strings.ToLower(right.Name)
	if leftName != rightName {
		if leftName < rightName {
			return -1
		}
		return 1
	}

	if left.Identity.Key < right.Identity.Key {
		return -1
	}
	if left.Identity.Key > right.Identity.Key {
		return 1
	}
	return 0
}
